content/blog/2024-10-21-testing-the-limits/index.md

```diff
@@ -0,0 +1,239 @@
+
+```bash 
+redis-benchmark -n 1000000 -t set,get -P 16 -q -a e41fb9818502071d592b36b99f63003019861dad -h 10.0.1.136
+SET: 394368.41 requests per second, p50=1.223 msec                    
+GET: 438058.53 requests per second, p50=1.135 msec 
+```
+
+We're up to 416,000 requests per second with a 1.2 msec (0.0012s) median latency. The mathmaticians out there might notice that this speed up is a lot more than the expected 47% increase. It's a 73% increase in requests per second and a 65% improvement in processing time. Whats happening?!
+
+While I haven't gone to machine code to verify here is what I believe is happening. The memory controller for the CM4 also lives on the CPU. Increasing the clock speed of the CPU also increases the performance of the memory. This means we almost are doubling the expected speedup by increasing the clock speed.
+
+## Adding IO Threading
+
+With all these gains I was super excited to try the new [IO threading available](https://valkey.io/blog/unlock-one-million-rps/) in Valkey 8. First we will take down the previous running docker instance with `docker compose -f valkey.yaml down`. Then we will modify the file to: 
+
+```yaml
+services:
+  valkey-1:
+    image: valkey/valkey:latest
+    hostname: valkey1
+    command: valkey-server --port 6379 --requirepass ${VALKEY_PASSWORD} --io-threads 5
+    volumes:
+      - ./data1:/data
+    network_mode: host
+
+volumes:
+  data1:
+    driver: local
+```
+
+Then we can `docker compose -f valkey.yaml up -d` to start it again. Remote into the benchmarking server to start the test and...?
+
+```bash 
+redis-benchmark -n 10000000 -t set,get -P 16 -q -a e41fb9818502071d592b36b99f63003019861dad -h 10.0.1.136
+SET: 345494.75 requests per second, p50=0.911 msec                    
+GET: 327858.09 requests per second, p50=0.879 msec  
+```
+
+Wait a second... These results are worse than the ones before? We went from 416k requests per second to 336k.... What's happening?
+
+![A picutre of the valkey server with 4 cores represented as boxes. In each of the four core boxes there are other boxes. 2 cores have 1 IO Thread Box, 1 Core has 2 IO thread boxes, and the last one has 1 IO Thread box along with a Valkey Process box.](images/io_threads.png)
+
+We have over-perscribed our CPU. This means we've created more worker threads than CPU cores. When a thread is under constant load it is competing with other threads on that core for resources. Not to mention they are also competing with the Valkey process for resources to read IO. 
+
+That's why [Valkey recommends](https://github.com/valkey-io/valkey/blob/a62d1f177b7888ec88035a0a1ce600fbc2280ce7/valkey.conf#L1337-L1341) setting the number of threads to be a value less than the number of cores you have. For our little 4 core server lets set it to be 2 threads and try again.
+
+```bash
+redis-benchmark -n 10000000 -t set,get -P 16 -q -a e41fb9818502071d592b36b99f63003019861dad -h 10.0.1.136
```

Comment:
Was this saturating the server, I doin't see the threads argument being passed in here. Without it, we can't saturate the target.

Reply:
Ah yeah, I ran all these in threaded mode on the PI. I'll go through and re-run the command and make sure the results are consistent.